### PR TITLE
Implement BulkUpdate with Celery

### PR DESCRIPTION
**Pull request by potsui**
@stvstnfrd @caseylitton PR for bulk update with Celery

NOTE: We need to remove commit 8f0b8fc (which I had added to enable celery locally) when we are ready to test on Jenkins.

To address in our f2f later today:
- Finish unit tests. Currently stuck at issue of getting the right `course_key_string` since the tester course returned by `CourseFactory` returns a deprecated version with slashes.
- Discovered while collecting performance data for poster: 
    - Largest course I was able to test on is Psychological Statistics, with 1173 problems. Celery task completed in 34s. Is this enough to warrant keeping as a Celery task?
    - On publish after update, a signal is sent out to trigger the task `update_search_index`, which takes significantly longer to complete than the update for large courses. This has no impact on our end as far as I can tell, but just something to be aware of.
- I ran the full `test_system` last night; no errors --- before, I think Jenkins tests were failing because I had lots of WIP code in the mix. Hopefully after we remove the commit noted above we won't have issues.
- @stvstnfrd mentioned yesterday he wanted to talk about changes to the UI status polling

Trello: https://trello.com/c/GB5DbAT9/552-add-ability-to-bulk-update-mvp-problem-settings

CC @jspayd @jlikhuva

### CODE REVIEW

File: cms/static/js/factories/bulkupdate.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
+define([
+    'domReady', 'js/views/bulkupdate', 'jquery', 'gettext', 'jquery.cookie'
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: See if we can remote jquery.cookie since we're not using it.

File: cms/static/js/factories/bulkupdate.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+        var validateData = function(data) {
+            var msg,
+                maxAttempts = data.maxAttempts,
+                showAnswer = data.showAnswer;
+
+            if (!maxAttempts && !showAnswer) {
+                return false;
+            }
+            if (maxAttempts) {
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Combine conditions

File: cms/static/js/factories/bulkupdate.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                if (maxAttempts < 0) {
+                    msg = gettext('Not a valid value for MaxAttempts. Please enter a different value.');
+                    $('.error-block').html(msg).show();
+                    console.error(msg);
+                    return false;
+                }
+            }
+            if (showAnswer) {
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Combine conditions

File: cms/static/js/factories/bulkupdate.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                if (SHOW_ANSWER_OPTIONS.indexOf(showAnswer) < 0) {
+                    msg = gettext('Not a valid value for ShowAnswer. Please enter a different value.');
+                    $('.error-block').html(msg).show();
+                    console.error(msg);
+                    return false;
+                }
+            }
+            return true;
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Add messages to an array.  If the msgs array has items return false.

File: cms/static/js/factories/bulkupdate.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                'answered',
+                'attempted',
+                'closed',
+                'finished',
+                'past_due',
+                'correct_or_past_due',
+                'never'
+            ];
+
+        var onComplete = function () {
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Change all 
```
var funcname = function()
TO
function funcname() {}
```

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            _save_request_status(request, session_status_string, -1)
+            return JsonResponse(
+                {
+                    'ErrMsg': 'Given value for show answer is not an available option',
+                    'Stage': -1
+                },
+                status=400,
+            )
+    _save_request_status(request, session_status_string, 2)
+    _update_advanced_settings(request, course_key_string, session_status_string, modified_settings)
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: add delay to run on celery.  Make sure order of the two celery tasks doesn't matter.

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    """
+    session_status = request.session.get('update_status')
+    if session_status is None:
+        session_status = request.session.setdefault("update_status", {})
+
+    session_status[key] = status
+    request.session.save()
+
+
+def _update_advanced_settings(request, course_key_string, session_status_string, modified_settings):
```
Comment (potsui) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Move to tasks.py

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    """
+    Internal method used to save to advanced settings
+    """
+    course_key = CourseKey.from_string(course_key_string)
+    course = modulestore().get_course(course_key, 3)
```
Comment (potsui) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Make store = modulestore() variable

File: cms/static/js/views/bulkupdate.js [hunk 9]
```diff
@@ -0,0 +1,342 @@
+/**
```
Comment (caseylitton) [inline cms/static/js/views/bulkupdate.js]: Clean up js UI based on conversation.

File: cms/static/js/factories/bulkupdate.js [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+                    data.maxAttempts,
+                    data.showAnswer,
+                    statusUrl.replace(
+                        'fillerMaxAttempts/fillerShowAnswer',
+                        data.maxAttempts + '/' + data.showAnswer
+                    )
+                );
+
+                BulkUpdate.pollStatus();
+                $submitBtn.prop('disabled', true);
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Move up to below validateData condition

File: cms/static/js/factories/bulkupdate.js [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+            // It could be that the user is simply refreshing the page
+            // so we need to be sure this is an actual error from the server
+            if (!unloading) {
+                $(window).off('beforeunload.update');
+            }
+            console.error('Error in making POST request', errMsg);
+        };
+
+        var onReset = function () {
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Might not need since js UI is being changed.

File: cms/templates/bulkupdate.html [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  </header>
+</div>
+
+<div class="wrapper-content wrapper">
+  <section class="content">
+    <article class="content-primary" role="main">
+      <section class="course-utility" id="bulk-update">
+        <form id="bulk-update-settings-input">
+          <header class="section-header">Problem Settings</header>
+          <p><b>Warning: </b>Any changes applied here will also modify the defaults for future problems in Advanced Settings.</p>
```
Comment (caseylitton) [inline cms/templates/bulkupdate.html]: use strong tag instead of b tag.

File: cms/templates/bulkupdate.html [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+      <p><b>Warning: </b>Public problems will automatically be republished. Please verify any drafts for public problems are ready to be published: <a href="${course_outline_url}">Course Outline</a></p>
+          <div id="max-attempts" class="setting-input-block">
+            <input type="checkbox" class="apply-checkbox" name="apply-checkbox" value="max-attempts" />
+            <h3 class = "setting-input-label" title="Max Attempts Setting">Max Attempts</h3>
```
Comment (caseylitton) [inline cms/templates/bulkupdate.html]: make 'class = ' into class= in the whole page

File: cms/templates/emails/utilities_bulk_update_done_message.txt [hunk 14]
```diff
@@ -0,0 +1,14 @@
+<%! from django.utils.translation import ugettext as _ %>
+${_("Hello, {username}").format(username=username)}
+
+% if success:
+${_("Your bulk update on the {course_name} course has successfully completed. Previously published problems have been automatically republished.").format(course_name=course_name)}
```
Comment (caseylitton) [inline cms/templates/emails/utilities_bulk_update_done_message.txt]: Check indents for conditional.  Make sure text itself is not then indented.  Check for syntax to fix or forget it and revert.

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+from edxmako.shortcuts import render_to_response
+from student.auth import has_course_author_access
+from util.json_request import JsonResponse
+from xmodule.modulestore.django import modulestore
+
+from opaque_keys.edx.keys import CourseKey
+
+from .tasks import bulk_update_problem_settings
+
+SHOW_ANSWER_OPTIONS = [
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Import from cms since we are already importing from cms.

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+import ddt
+
+from xmodule.modulestore.tests.factories import CourseFactory
+
+from opaque_keys.edx.keys import CourseKey
+
+from cms.djangoapps.contentstore.tests.utils import CourseTestCase
+from contentstore.utils import reverse_course_url
+
+from ..views.utilities.bulkupdate import utility_bulkupdate_handler, SHOW_ANSWER_OPTIONS
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: import from cms too

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+    to use when getting and setting the session status
+    """
+    return 'course_key_string={}&max_attempts={}&show_answer={}'.format(
+        course_key_string,
+        max_attempts,
+        show_answer,
+    )
+
+
+def _reverse_course_url(handler, course_key_string, kwargs=None):
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: maybe import from cms too.

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+def _utility_bulkupdate_post_handler(request, course_key_string):
+    """
+    Internal bulkupdate handler for POST operation
+    """
+    modified_settings = {}
+    user = request.user
+
+    # Retrieve settings from request
+    try:
+        max_attempts = request.POST.get('maxAttempts')
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Check types for max_attempts and show_answer.  Do we need to cast them down below?

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+                'ErrMsg': 'Missing or invalid arguments',
+                'Stage': -1
+            },
+            status=400,
+        )
+
+    # Validate settings
+    _save_request_status(request, session_status_string, 1)
+
+    if max_attempts and max_attempts != 'null':
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Maybe with hold from data if it is null.

File: cms/static/js/factories/bulkupdate.js [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+            $submitBtn = $('.view-bulkupdate .submit-button'),
+            DEFAULT_ERROR_MESSAGES = [
+                gettext('There was an error while sending your request.') + '\n',
+                gettext('There was an error while validating the setting values you submitted.') + '\n',
+                gettext('There was an error while updating your advanced settings.') + '\n',
+                gettext('There was an error while updating your problem settings in our database.') + '\n'
+            ],
+            unloading = false,
+            previousUpdate = BulkUpdate.storedUpdate(),
+            SHOW_ANSWER_OPTIONS = [
```
Comment (caseylitton) [inline cms/static/js/factories/bulkupdate.js]: Possibly pass in through html.  Lower priority.

File: cms/static/js/factories/bulkupdate.js [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+            $maxAttempts.val(defaultMaxAttempts);
+            $showAnswer.val(defaultShowAnswer);
+        };
+
+        $(window).on('beforeunload', function () { unloading = true; });
+
+        var getSettingsData = function() {
+            var data = {};
+            if ($applyMaxAttempts.is(':checked')) {
+                data.maxAttempts = $maxAttempts.val();
```
Comment (caesar2164) [inline cms/static/js/factories/bulkupdate.js]: cast this as an int

File: cms/static/js/factories/bulkupdate.js [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+            } else {
+                data.maxAttempts = null;
+            }
+            if ($applyShowAnswer.is(':checked')) {
+                data.showAnswer = $showAnswer.val();
+            } else {
+                data.showAnswer = null;
```
Comment (caesar2164) [inline cms/static/js/factories/bulkupdate.js]: make this an empty string

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+
+    Currently updates max_attempts and showanswer for all problems in a
+    given course and sets as settings for future problems in advanced
+    settings
+    """
+    course_key = CourseKey.from_string(course_key_string)
+    if not has_course_author_access(request.user, course_key):
+        raise PermissionDenied()
+
+    if 'text/html' in request.META.get('HTTP_ACCEPT', 'text/html'):
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Maybe only check for 'GET' or 'POST' only.

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        """
+        super(BulkUpdateTest, self).setUp()
+        self.course = CourseFactory.create()
+        self.bulkupdate_url = reverse_course_url('utility_bulkupdate_handler', self.course.id)
+
+    def check_modified_advanced_settings(self, settings):
+        """
+        Helper for test cases to check advanced advanced settings have been modified
+        """
+        if settings['maxAttempts'] != 'null':
```
Comment (caesar2164) [inline openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: this should be the element `null` not string

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py (hunk 19 above)
Comment (caesar2164) [reply openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: if `max_attempts` is defined as an int or `null` this can just be:
```python
if max_attempts:
```

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+                {
+                    'ErrMsg': 'Given value for max attempts is negative',
+                    'Stage': -1
+                },
+                status=400,
+            )
+        else:
+            modified_settings['max_attempts'] = max_attempts
+
+    if show_answer and show_answer != 'null':
```
Comment (caesar2164) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: if we default this to `''` this condition can become:
```python
if show_answer:
```

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/tasks.py [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+    store = modulestore()
+    problems = _get_course_problems(course_key)
+    with store.bulk_operations(course_key):
+        for problem in problems:
+            for metadata_key, value in metadata.items():
+                field = problem.fields[metadata_key]
+                try:
+                    value = field.from_json(value)
+                except Exception as exception:
+                    LOGGER.error(exception)
```
Comment (caseylitton) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/tasks.py]: maybe continue if exception or add write_to to try else

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py (hunk 15 above)
Comment (potsui) [reply openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Ended up building from a `capa_base.py` file in `common`

File: cms/templates/emails/utilities_bulk_update_done_message.txt (hunk 14 above)
Comment (potsui) [reply cms/templates/emails/utilities_bulk_update_done_message.txt]: Does make text indented, quickly looked for solution, and just reverted change. Added item to follow-up work checklist in Trello card

**Comment by potsui** [comment]
@stvstnfrd @caseylitton got everything except indentation (noted above and in Trello) since it's low priority. Unit testing still incomplete because of issue we discussed f2f

File: pavelib/servers.py [hunk 27]
```diff
@@ -157,7 +157,7 @@ def celery(options):
     Runs Celery workers.
     """
     settings = getattr(options, 'settings', 'dev_with_worker')
-    run_process(django_cmd('lms', settings, 'celery', 'worker', '--beat', '--loglevel=INFO', '--pythonpath=.'))
+    run_process(django_cmd('studio', settings, 'celery', 'worker', '--beat', '--loglevel=INFO', '--pythonpath=.'))
```
Comment (stvstnfrd) [inline pavelib/servers.py]: Do we need/want both of these lines?
How can/would the two sets of workers run simultaneously?

File: cms/static/cms/js/build.js [hunk 28]
```diff
@@ -34,6 +34,7 @@
         modules: getModulesList([
             'js/factories/asset_index',
             'js/factories/base',
+            'js/factories/bulkupdate',
```
Comment (stvstnfrd) [inline cms/static/cms/js/build.js]: Hrmm, is there no corresponding entry for `bulksettings` or the other utilities?

File: cms/static/js/factories/bulkupdate.js [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+            $submitBtn = $('.view-bulkupdate .submit-button'),
+            $errorMessage = $('.view-bulkupdate .server-message-wrapper .error-server-message'),
+            $successMessage = $('.view-bulkupdate .server-message-wrapper .success-server-message'),
+            SHOW_ANSWER_OPTIONS = [],
+            VALIDATION_ERROR_MESSAGES = {
+                maxAttempts: gettext('MaxAttempts must be a non-negative integer. Please enter a different value.'),
+                showAnswer: gettext('Not a valid value for showAnswer. Please enter a different value.'),
+            };
+
+        // Fill SHOW_ANSWER_OPTIONS with option values from the HTML
```
Comment (stvstnfrd) [inline cms/static/js/factories/bulkupdate.js]: Nice!

File: cms/static/sass/views/_bulkupdate.scss [hunk 30]
```diff
@@ -0,0 +1,89 @@
+// Studio - Course Utilities - BulkUpdate
+// ====================
+
+.view-bulkupdate {
+
+  .content-primary, .content-supplementary {
```
Comment (stvstnfrd) [inline cms/static/sass/views/_bulkupdate.scss]: Do you know if there are any common/shared CSS files we could include that already provide this basic styling?

File: cms/templates/bulkupdate.html [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+  </div>
+
+  <div class="wrapper-content wrapper">
+    <section class="content">
+      <article class="content-primary" role="main">
+        <section class="course-utility" id="bulk-update">
+          <form id="bulk-update-settings-input">
+            <header class="section-header">Problem Settings</header>
+              <p>
+                <strong>Warning: </strong>
```
Comment (stvstnfrd) [inline cms/templates/bulkupdate.html]: Probably too late now, but for followup, the strings in this template should be i18n'ed.

File: lms/envs/devstack.py [hunk 32]
```diff
@@ -16,7 +16,8 @@
 SITE_NAME = 'localhost:8000'
 PLATFORM_NAME = ENV_TOKENS.get('PLATFORM_NAME', 'Devstack')
 # By default don't use a worker, execute tasks as if they were local functions
-CELERY_ALWAYS_EAGER = True
+CELERY_ALWAYS_EAGER = False
+FEATURES['ENABLE_SERVICE_STATUS'] = True
```
Comment (stvstnfrd) [inline lms/envs/devstack.py]: This will be reverted, right?

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 33]
```diff
@@ -0,0 +1,94 @@
+"""
+Unit tests for bulk update problem settings utility
+
+TODO Write unit tests for bulk_update_problem_settings from utilities.tasks
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: Did you manage to get the tests cleaned up?

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+
+from opaque_keys.edx.keys import CourseKey
+
+from cms.djangoapps.contentstore.utils import reverse_course_url
+
+from .tasks import bulk_update_problem_settings
+
+SHOW_ANSWER_OPTIONS = []
+for value in CapaFields.__dict__['showanswer'].values:
+    SHOW_ANSWER_OPTIONS.append(value['value'])
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Not a dealbreaker, but the more Pythonic approach here would be to use a list comprehension:

```python
SHOW_ANSWER_OPTIONS = [
    value['value']
    for value in CapaFields.__dict__['showanswer'].values
]
```

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py [hunk 35]
```diff
@@ -0,0 +1,94 @@
...
+from xmodule.capa_base import CapaFields
+from xmodule.modulestore.tests.factories import CourseFactory
+
+from cms.djangoapps.contentstore.tests.utils import CourseTestCase
+from contentstore.utils import reverse_course_url
+
+from ..views.utilities.bulkupdate import SHOW_ANSWER_OPTIONS
+
+
+SHOW_ANSWER_OPTIONS = []
```
Comment (stvstnfrd) [inline openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: We shouldn't be rebuilding this here,
since we're importing it already, right?

Comment (potsui) [reply openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: Forgot to take the import out, but now that it's brought up...for tests should we be rebuilding the list or importing from the file being tested?

File: lms/envs/devstack.py (hunk 32 above)
Comment (potsui) [reply lms/envs/devstack.py]: Yes, the first commit of this PR needs to be reverted

File: cms/envs/devstack.py [hunk 36]
```diff
@@ -57,7 +57,8 @@
 ################################# CELERY ######################################
 
 # By default don't use a worker, execute tasks as if they were local functions
-CELERY_ALWAYS_EAGER = True
+CELERY_ALWAYS_EAGER = False
```
[the remark this answers is not in the document]
Comment (potsui) [reply cms/envs/devstack.py]: Yes

File: cms/static/cms/js/build.js (hunk 28 above)
Comment (potsui) [reply cms/static/cms/js/build.js]: I don't see any

File: openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py (hunk 33 above)
Comment (potsui) [reply openedx/stanford/cms/djangoapps/contentstore/tests/test_bulkupdate.py]: Resolved the issue from our convo f2f (used `@mock.patch` as a workaround) --- I've written tests for checking advanced settings and problem settings are modified, but not edge cases, exceptions, etc., yet

File: pavelib/servers.py (hunk 27 above)
Comment (potsui) [reply pavelib/servers.py]: I haven't looked into this much. The approach in edx docs _is_ scalable by running workers in separate terminals, but I wasn't able to spin up celery their way: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/environment/celery.html

Without knowing anything else on the subject, I think we would at least probably want separate commands / flags for specifying which worker to run rather than keeping both lines and spinning them both up.

File: openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py (hunk 34 above)
Comment (potsui) [reply openedx/stanford/cms/djangoapps/contentstore/views/utilities/bulkupdate.py]: Done

**Comment by potsui** [comment]
PROTIP: Publishing the course emits a signal to run other celery tasks (i.e. `update_search_index`) which takes longer and may hold up your celery worker when developing on devstack with one worker. My workaround for developing quickly is using flower to view the task queue and terminating outstanding ones. You can also set rate limits per task type on Flower.

**Comment by potsui** [comment]
Devstack setup notes
---

Run Celery Workers on Devstack (CMS)

1. In `pavelib/servers.py` in `def celery` change `run_process(django_cmd('lms', settings, 'celery', 'worker', '--beat', '--loglevel=INFO', '--pythonpath=.’))` to use studio instead
2. Set `CELERY_ALWAYS_EAGER = False` in `cms/envs/devstack.py` (By default we don’t use a worker, execute tasks as if they were local functions)
3. Start server, `$paver devstack studio`
4. Open 2nd terminal on box as root user
5. Check status of rabbitmq, `$sudo rabbitmqctl status`

rabbitmqstatus

6. `$sudo su edxapp`
7. Start celery workers in edx-platform, `$paver celery --settings=devstack`
8. The task should run in the celery terminal as opposed to the server terminal.  

FYI - Seeing celery workers status on server
1. Add `FEATURES[‘ENABLE_SERVICE_STATUS’] = True` to `devstack.py`
2. Once `$paver devstack studio` is running you can use http://localhost:8001/status/celery/ to see celery info

FYI - Flower - Drop in UI to monitor celery tasks that can actually terminate worker/terminate tasks.  
1. Third terminal window on box
2. `$sudo su edxapp`
3. `pip install flower`
4. `celery flower --broker=amqp://celery:celery@localhost/ --port=8001`

**Comment by potsui** [comment]
Current UI
new ui

**Comment by stvstnfrd** [comment]
Closing in favor of: https://github.com/Stanford-Online/edx-platform/pull/895
